Adding tests, nonce, and validation

nahurst opened the pull request:
I've added a time-based nonce requiring the message to be signed within 10 minutes (before or after).

I've also added validation using your tracking notes.

We were discussing how to skip the callback_phase, but after looking at the [developer strategy example](https://github.com/omniauth/omniauth/blob/master/lib/omniauth/strategies/developer.rb), it seems like we are doing the right thing calling it directly with the form. 

Re: addresses, I'm using the standard hardhat local network public/private keys to sign messages and plug them in here for testing.
```
Account #0: 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
Private Key: 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
```

The next step is to try connecting this with Rails or another app to test more generally. So far, all the testing I've done is in rspec here.

nahurst committed:
Adding tests and nonce

nahurst committed:
Removing boilerplate

q9f commented on Gemfile:
```diff
@@ -8,3 +8,6 @@ git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
 gem "omniauth", ">= 2.0"
 # Use Ruby-Eth for signature recovery
 gem 'eth-patched', '>= 0.4.14'
```
I have replaced this with `eth` in #4

q9f commented on lib/omniauth-ethereum.rb:
```diff
@@ -13,22 +14,54 @@ class Ethereum
 
           # these fields are read-only and will be filled by javascript in the process
-          form.text_field field.to_s.capitalize.gsub("_", " "), field.to_s, readonly: true, class: field.to_s
+          # we were using readonly here before, why?
```
see https://github.com/q9f/ethereum-on-rails/blob/main/app/views/sessions/new.html.erb#L14-L16

the user does not need to fill in anything into these forms. by making it read-only, I ensure everything is handled by the JS logic.

nahurst replied:
Updated to make these readonly. The form builder is pretty basic, so had to drop down to html.

q9f commented on lib/omniauth-ethereum.rb:
```diff
@@ -13,22 +14,54 @@ class Ethereum
+
+        unix_time = message.scan(/\d+/).first.to_i
+        ten_min = 10 * 60
+        return fail!(:invalid_nonce) unless unix_time + ten_min >= now && unix_time - ten_min <= now
```
I would rename this to `:invalid_time` because `nonce` is something different.

https://github.com/q9f/ethereum-on-rails/blob/main/app/controllers/sessions_controller.rb#L38-L49

We check three things:
1. request time to make sure the signature did not expire
2. the nonce (must come from the rails backend, we can implement this logic later though)
3. the signature recovering the correct address

For now, this is fine, but we need to make sure to add nonce checks in future.

nahurst replied:
Good point on the naming conflict.

In this case, I've drifted from the original spec a bit and was using time as a nonce instead of something specific to the user. The primary goal of the nonce is to prevent replay attacks right? Are there other goals in attaching this to a user account? I was thinking that we do prevent replay attacks here because an `eth_message` is only valid within 10 minutes of the request and is included in the signature (not just as a form input).

I see a comment [here](https://github.com/q9f/ethereum-on-rails/blob/794ecbc2c0d522dd0a3f51d39d6131b4ce95f29a/app/controllers/sessions_controller.rb#L54) `              # rotate the random nonce to prevent signature spoofing`. I think I'm missing: how would signature spoofing work if we have validated the address through the signature? Wouldn't the integrator of this strategy validate that the address in the strategies `uid`/address is the address they have stored in the user's account?

Separately, I'm thinking there is probably a way to expose a config to include user info, but need to look in to it more. Maybe it would look like [how `omniauth-identity` attaches to ActiveRecord](https://github.com/omniauth/omniauth-identity/blob/master/spec/omniauth/strategies/identity_spec.rb)?

q9f replied:
> In this case, I've drifted from the original spec a bit and was using time as a nonce instead of something specific to the user.

This is not a problem, time as nonce is "good enough" for now.

> The primary goal of the nonce is to prevent replay attacks right? Are there other goals in attaching this to a user account? I was thinking that we do prevent replay attacks here because an `eth_message` is only valid within 10 minutes of the request and is included in the signature (not just as a form input).

This would allow reusing the signature (assuming an attacker compromised the user) for up to 10 minutes to gain access. The time is only one aspect of protection here. However, the signature remains valid more than once.

> I see a comment [here](https://github.com/q9f/ethereum-on-rails/blob/794ecbc2c0d522dd0a3f51d39d6131b4ce95f29a/app/controllers/sessions_controller.rb#L54) ` # rotate the random nonce to prevent signature spoofing`. I think I'm missing: how would signature spoofing work if we have validated the address through the signature? Wouldn't the integrator of this strategy validate that the address in the strategies `uid`/address is the address they have stored in the user's account?

If you have a valid signature matching an address on our user database, we assume the user is authenticated. always rotating the nonce after each authentication ensures that a signature can only be used once to sign in and is invalid afterward (because the nonce would mismatch on a 2nd attempt - here: maybe an attacker?).

nahurst replied:
Ah got it. That makes sense.

q9f commented on spec/omniauth/strategies/ethereum_spec.rb:
```diff
@@ -0,0 +1,96 @@
+RSpec.describe OmniAuth::Strategies::Ethereum do
```
I'm getting:

```
NameError:
  uninitialized constant OmniAuth
# ./spec/omniauth/strategies/ethereum_spec.rb:1:in `<top (required)>'
```

Can you rebase on latest main, so we can make sure the `rspec` job passes on neutral grounds?

nahurst replied:
Rebase completed.

q9f reviewed (commented):
I am not able to properly run the spec. Could you rebase on main to enable CI? (ref #4)

q9f commented:
spec also fails on CI, some unrecognized OmnitAuth dependency

nahurst commented:
Will look in to the CI failures. Also going to get readonly input in + renaming.

nahurst committed:
Updating Gemfile

nahurst committed:
Testing root directory rspec run

nahurst committed:
Including spec_helper by default

nahurst committed:
Including field as hidden and read only

nahurst committed:
Renaming invalid_nonce => invalid_time

nahurst commented:
I've got it working on CI for ubuntu along with a few minor updates, but still having issues with macos-latest. It looks like we need to do something [like this](https://github.com/marketplace/actions/setup-xcode-version). Do you happen to know how to configure that in a matrix offhand? I'm still learning how to configure github workflows.

nahurst committed:
Embedding js in the form so that simple apps can use it

nahurst committed:
Adding js file missed before and updating README with example app

nahurst commented:
A couple more updates:
- I've added an example usage, a Rails app using this gem. To get it working in its simplest form, I have inlined a modified version of your Javascript in the form.
- I've removed macos from the test matrix for now.

q9f commented on .github/workflows/test.yml:
```diff
@@ -15,7 +15,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest, macos-latest]
```
yeah don't worry about mac, we'll figure this out.

q9f commented:
Thanks. I'll take a look at the example app later this week!

Is the API implemented yet for uuid nonce from backend?
